doc/man/man3/Alaris.3
.TH "Alaris" 3 "25 May 2010" "Version 1.6" "Alaris IRC Bot" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Alaris \- 
.SH SYNOPSIS
.br
.PP
.SS "Packages"

.in +1c
.ti -1c
.RI "package \fBCore\fP"
.br
.ti -1c
.RI "package \fBExtras\fP"
.br
.ti -1c
.RI "package \fBIrc\fP"
.br
.ti -1c
.RI "package \fBLuaPlugin\fP"
.br
.ti -1c
.RI "package \fBMainPlugin\fP"
.br
.ti -1c
.RI "package \fBMangosRssPlugin\fP"
.br
.in -1c
.SH "Author"
.PP 
Generated automatically by Doxygen for Alaris IRC Bot from the source code.


Atom/CHANGES.txt
15 Jul 2004 (0.4.3 release):
	- removed xml:base from TODO list
	- fixed some doc

14 Jul 2004:
	- fixed a bug in media types parsing
	- enhanced writing capabilities
	- removed some code duplication
	- fixed a bug in AtomPersonConstruct parsing
	- added xml:base support
	- added a test feed for xml:base capabilities
	- tests\Atom.Core\AtomContentTest.cs: renamed a test
	- tests\Atom.Core\AtomContentTest.cs: added TestXmlBaseContentUri to test presence of xml:base attribute in certain contents
	- tests\Atom.Core\AtomEntryTest.cs: added TestNameIsNotEmail to ensure that <name> tags are not emails
	- tests\Atom.Core\AtomFeedTest.cs: added TestXmlBaseFeedUri to test the presence of xml:base attribute in certain feeds
	- tests\Atom.Core\AtomFeedTest.cs: added TestNameIsNotEmail to ensure that <name> tags are not emails

08 Jul 2004:
	- Atom.Utils\Utils.cs: added a public GetVersion() method to obtain the version of the assembly
	- Atom.Utils\DefaultValues: fixed GeneratorVersion internal string
	- tests\Atom.Utils\UtilsTest.cs: added a test for GetVersion() method

03 Jul 2004:
	- added XPathNavigatorReader.cs (part of mvp-xml project [http://sourceforge.net/projects/mvp-xml]
	    (many thanks to Daniel Cazzulino [http://weblogs.asp.net/cazzu]
	- updated nant build file

29 Jun 2004 (0.4.2 internal release):
	- fixed new Uri properties
	- changed some stuff according to FxCop rules
	- Atom.Core\AtomFeed.cs: changed Save(HttpWebResponse) to Save(WebResponse)
	- Atom.Core\AtomFeed.cs: changed Load(HttpWebRequest) to Load(WebRequest)
	- Atom.Core\AtomFeed.cs: added Load(System.Uri) overload
	- Atom.Core\AtomEntry.cs: changed Save(HttpWebResponse) to Save(WebResponse)
	- Atom.Core\AtomEntry.cs: changed Load(HttpWebRequest) to Load(WebRequest)
	- Atom.Core\AtomEntry.cs: added Load(System.Uri) overload
	- I'm ready to integrate mvp-xml library
	- added AtomFeedTest.TestUri and AtomFeedTest.TestUri

24 Jun 2004:
	- Atom.Utils\DefaultValues.cs: fixed generator version
	- Atom.Core\AtomEntry.cs: added a Uri property for an easy identification of the uri of the entry
	- Atom.Core\AtomEntry.cs: fixed documentation
	- Atom.Core\AtomFeed.cs: removed etag internal private variable

23 Jun 2004:
	- AtomFeed.cs: added a Uri property for an easy identification of the uri of the feed
	- AtomFeed.cs: fixed documentation
	- removed filtering based on file extensions
	- renamed AtomFeedTest.TestUri to TestSaveToFile

21 Jun 2004:
	- fixed AtomPersonConstruct.cs documentation

13 May 2004 (0.4.1 release):
	- bugfix relase

29 Mar 2004:
	- fixed a bug in the serialization of AtomCreated tag

26 Feb 2004:
	- added quite every media type out there in Enum.MediaType
	- embedded mediatypes.txt as resource
	- rewritten the media type parsing stuff to gain advantage of the embedded resource
	- fixed nant build file and makefile

25 Feb 2004:
	- added Load(XmlReader) and Save(XmlWriter) to AtomEntry and AtomFeed
	- added tests for XmlReader based Load methods (AtomEntry/AtomFeed)
	- added tests for XmlWriter based Save methods (AtomEntry/AtomFeed)
	
24 Feb 2004 (0.4 Release):
	- updated and tested Makefile
	- updated the doc
	- updated the webpage

23 Feb 2004:
   * pre release state
	- major architecture rework:
		- collapsed Atom.Core.BaseClasses into Atom.Core
		- removed AtomAuthor, AtomLinkConstruct, AtomContributor,
		  all content construct classes except for AtomContent,
		  removed ExceptionCollection
	- exposed the internal buffer to an internal property
	- merged Load/Save test methods
	- added test for invalid feeds/entries

20 Feb 2004:
	- fixed escape and unescape in utils

19 Feb 2004:
	- added localname, namespace uri, namespace prefix for every element
	- implemented a base class for every additional element
	- added DublinCore elements support
	
14 Feb 2004:
	- added a skeleton for extendibility issues
	- added tests for Utils classes

13 Feb 2004
	- fixed AtomReader to parse namespaced atom elements
	- fixed a bug with html elements inside atom tags
	- xml:lang attribute at feed level has become optional
	- added xml:lang attribute for every element
	- fixed a bug in base64 utils function

12 Feb 2004
	- added an Encoding property to AtomFeed to specifying content encoding.
	- AtomFeed now saves themselves (against a path, a stream, a TextWriter or an HttpWebResponse)
	- AtomFeed class now haves overloaded Load() static methods (to load from a uri, a stream, a TextReader or an HttpWebRequest)
	- AtomWriter and AtomReader are no longer exposed and cannot be used.
	- moved enum values from Atom.Utils to Atom.Core namespace

11 Feb 2004
	- added new tests for Save, Load and BinarySerialization capability of AtomFeed and AtomEntry

08 Feb 2004
	- fixed a bug in parsing of namespaced scoped elements
	- fixed the feed id compliancy issue in writing
	- marked all Atom.Core classes as Serializable (so the state of the class could be serialized)
	- DefaultValues class is no longer istantiable or inheritable
	- added a StringBuilder buffer for every Atom.Core class
	- added the serialization capability for every class in the Atom.Core namespace
	- every class in the Atom.Core namespace now can be converted to its string representation separately 

07 Feb 2004
	- added Makefile (thanks to John Luke of the Mono Project)
    - replaced Atom.NET.Release.xml with Atom.NET.xml
	- replaced Util with explicit namespace in AtomReader
	- added a nant build file (see http://nant.sf.net for more infos)

03 Feb 2004
	* 0.3 release
	* CHANGELOG:
		- fixed AtomGenerator constructors
		- written AtomReader to parse atom feeds
		- fixed a bug in a private method that invalidated parsing if the
			first <link> hasn't "alternate" attribute
		- added IsIso8601Date(), IsIso8601DateLocal(), IsIso8601DateTZ() to validate iso 8601 date times.
		- fixed some bugs with date/time representation in <modified>, <created>, <issued> elements
		- fixed a bug in the serialization of xml:lang attribute

alaris/Alaris.cs
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Diagnostics;
using ICSharpCode.SharpZipLib;
using Mono.Math;
using System.Reflection;
using System.Threading;
using System.Timers;
using Alaris.Irc;
using Alaris.Extras;
using System.Net.Sockets;


namespace Alaris.Core
{
	/// <summary>
	/// The main class for Alaris.
	/// </summary>
	public partial class AlarisBot : IThreadContext
	{
		private Connection _connection;
		private readonly ScriptManager _manager;
		private string _nick = "alaris";
		private string _server = "irc.rizon.net";
		private bool _confdone = false;
		private bool _nickserv = false;
		private string _nspw = "";
		private List<string> _channels = new List<string>();
		private string _anick, _auser, _ahost;
		private CrashHandler sCrashHandler = Singleton<CrashHandler>.Instance;
		private DatabaseManager sDatabaseManager = Singleton<DatabaseManager>.Instance;
		private readonly Guid _guid = Guid.NewGuid();
		private readonly string _configfile;
		private const int listener_port = 35221;
		private const string ACS_HOST = "127.0.0.1";
		private const int ACS_PORT= 35220;
		/// <summary>
		/// MySQL support enabled or not.
		/// </summary>
		public bool MysqlEnabled = false;
		/// <summary>
		/// MySQL data (host, user etc.).
		/// Size: 4 (DB is last)
		/// </summary>
		public string[] MysqlData = new string[4];
		
		/// <summary>
		/// Determines whether the communication to and dependance of alaris_server is set.
		/// </summary>
		public readonly static bool AlarisServer = false;
		
		/// <summary>
		/// The acs_rand_request_channel.
		/// </summary>
		public string acs_rand_request_channel;
		
		/// <summary>
		/// Gets or sets the thread pool.
		/// </summary>
		/// <value>
		/// The thread pool.
		/// </value>
		public CThreadPool Pool { get; private set; }
		
		private AlarisBot() : this("alaris.conf") {}
		
		/// <summary>
		/// Creates a new instacne of Alaris bot.
		/// </summary>
		private AlarisBot (string config)
		{
			Log.Notice("Alaris", "Initalizing...");
			_configfile = config;
			var cargs = new ConnectionArgs(_nick, _server);
			Log.Debug("Identd", "Starting service...");
			Identd.Start(_nick);
			Log.Success("Identd", "Service daemon running.");
			Log.Notice("Alaris", "Setting up connection...");
			Pool = new CThreadPool(4);
			
			_connection = new Connection(cargs, true, false);
			var responder = new CtcpResponder(_connection);
			responder.VersionResponse = "Alaris " + Utilities.GetBotVersion();
			responder.SourceResponse = "http://www.wowemuf.org";
			responder.UserInfoResponse = "Alaris multi-functional bot.";
			
			_connection.CtcpResponder = responder;
			Log.Notice("Alaris", "Text encoding: UTF-8");
			_connection.TextEncoding = Encoding.GetEncoding("Latin1");
			
			Log.Success("CTCP", "Enabled.");
			
			Log.Notice("ScriptManager", "Initalizing...");
			_manager = new ScriptManager(ref _connection, ref _channels);
			//_manager.LoadPlugins();
			//Thread.Sleep(2000);
			//Log.Success("ScriptManager", "Setup complete");
			
			Pool.Enqueue(_manager);
			
			SetupHandlers();
			
		}
		
		/// <summary>
		/// Gets the listener port.
		/// </summary>
		/// <returns>
		/// The listener port.
		/// </returns>
		public int GetListenerPort() { return listener_port; }
		
		/// <summary>
		/// Releases unmanaged resources and performs other cleanup operations before the <see cref="Alaris.Core.AlarisBot"/>
		/// is reclaimed by garbage collection.
		/// </summary>
		~AlarisBot()
		{
			Log.Debug("Alaris", "~AlarisBot()");
		}
		
		/// <summary>
		/// Gets the GUID.
		/// </summary>
		public Guid GetGuid() { return _guid; }
		
		/// <summary>
		/// Run this instance.
		/// </summary>
		public void Run()
		{
			var ls = new List<Exception>();
			sCrashHandler.HandleReadConfig(ReadConfig, _configfile, ref ls);
			ls.Clear();
			
			// start database server.
			if(MysqlEnabled)
			{	
				sDatabaseManager.Initialize(MysqlData[0], MysqlData[1], MysqlData[2], MysqlData[3]);
			}
			
			Connect();
		}
		
		/// <summary>
		/// Gets the class' current crash handler.
		/// </summary>
		/// <returns>
		/// The crash handler.
		/// </returns>
		public CrashHandler GetCrashHandler() { return sCrashHandler; }
		
		private void SetupHandlers()
		{
			Log.Notice("ScriptManager", "Setting up event handlers.");
			_manager.RegisterOnRegisteredHook(OnRegistered);
			_manager.RegisterOnPublicHook(OnPublicMessage);
			_connection.CtcpListener.OnCtcpRequest += OnCtcpRequest;
			Log.Success("ScriptManager", "Event handlers are properly setup.");
		}
		
		
		
		/// <summary>
		/// Reads and parses the specified config file.
		/// </summary>
		/// <param name="configfile">
		/// The config file name.
		/// </param>
		public void ReadConfig(string configfile)
		{
			if(!File.Exists("./" + configfile))
				throw new FileNotFoundException("The config file specified could not be found. It is essential to have a configuration file in the directory of the bot.");
			
			// read conf file.
			Log.Notice("Config", "Reading configuration file: " + configfile);
			var reader = new StreamReader("./"+configfile);
			string config = reader.ReadToEnd();
			reader.Close();
			reader.Dispose();
			
			
			var serverRegex = new Regex(@"server\s=\s(?<server>\S+)");
			if(!serverRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			_server = serverRegex.Match(config).Groups["server"].ToString();
			
			var nickRegex = new Regex(@"nick\s=\s(?<nick>\S+)");
			if(!nickRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			_nick = nickRegex.Match(config).Groups["nick"].ToString();
			
			
			var nickservRegex = new Regex(@"nickserv\s=\s(?<nickserv>\S+)");
			
			if(!nickservRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			_nspw = nickservRegex.Match(config).Groups["nickserv"].ToString();
			_nickserv = (_nspw == "nothing") ? false : true;
			
			var chanRegex = new Regex(@"channels\s=\s(?<channels>\S+)");
			
			if(!chanRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			string chn = chanRegex.Match(config).Groups["channels"].ToString();
			string[] chns = chn.Split(',');
			
			foreach(string cs in chns)
				if(!string.IsNullOrEmpty(cs))
					_channels.Add(cs);
			
			
			var admRegex = new Regex(@"admin_data\s=\s(?<adm>\S+)");
			if(!admRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			string[] adms = admRegex.Match(config).Groups["adm"].ToString().Split(',');
			_auser = adms[0];
			_anick = adms[1];
			_ahost = adms[2];
			
			_confdone = true;
			Utilities.AdminUser = _auser;
			Utilities.AdminNick = _anick;
			Utilities.AdminHost = _ahost;
			
			var mysqlRegex = new Regex(@"mysql_enabled\s=\s(?<enabled>\d)\s*\n*\r*mysql_data=(?<host>\S+),(?<user>\S+),(?<pass>\S+),(?<db>\S+)");
			
			if(!mysqlRegex.IsMatch(config))
				throw new ConfigFileInvalidException("The specified configuration file is invalid.");
			
			var msmatch = mysqlRegex.Match(config);
			MysqlEnabled = (msmatch.Groups["enabled"].ToString() == "1");
			
			if(MysqlEnabled)
			{
				Log.Debug("Alaris", "MySQL support is ON.");
				MysqlData[0] = msmatch.Groups["host"].ToString();
				MysqlData[1] = msmatch.Groups["user"].ToString();
				MysqlData[2] = msmatch.Groups["pass"].ToString();
				MysqlData[3] = msmatch.Groups["db"].ToString();
			}
			else
				Log.Debug("Alaris", "MySQL support is OFF.");
			
			
			Log.Success("Config", "File read and validated successfully.");

			
		}
		
		/// <summary>
		/// Sends the packet to ACS.
		/// </summary>
		/// <param name='packet'>
		/// Packet.
		/// </param>
		public void SendPacketToACS(AlarisPacket packet)
		{
			if(!AlarisServer)
				return;
			
			var client = new TcpClient();
			try
			{
				
				var endp = new IPEndPoint(IPAddress.Parse(ACS_HOST), ACS_PORT);
				client.Connect(endp);
				
				Thread.Sleep(300);
				
				if(client.Connected)
				{
					
					var stream = client.GetStream();
					Log.Debug("AlarisServer", "Connected. Sending packet.");
	
					var encoder = new UTF8Encoding();
					
					byte[] buffer = encoder.GetBytes(packet.GetNetMessage());
					
					stream.Write(buffer, 0, buffer.Length);
					stream.Flush();
					
					Log.Debug("AlarisServer", "Packet sent.");
					
					stream.Close();
					
					client.Close();
					
					Log.Debug("AlarisServer", "Connection closed.");
				}
			}
			catch
			{
				Log.Error("Alaris", "Couldn't send ACS packet.");
			}
			finally
			{
				client.Close();
			}
		}
		
		/// <summary>
		/// Establishes the connection to the previously specified server.
		/// </summary>
		public void Connect()
		{
			if(!_confdone)
				throw new Exception("The config file has not been read before connecting.");
			
			Log.Notice("Alaris", "Establishing connection...");
			try { _connection.Connect(); }
			catch(Exception x) { Log.Error("Alaris", x.Message); Identd.Stop(); }
		}
		
		/// <summary>
		/// Disconnects the bot from the IRC server.
		/// </summary>
		/// <param name="rsr">
		/// Reason for disconnect.
		/// </param>
		public void Disconnect(string rsr)
		{
			Log.Notice("Alaris", "Disconnecting...");
			Pool.Free();
			
			if(Identd.IsRunning())
			{
				Identd.Stop();
				Log.Success("Identd", "Stopped service daemon");
			}
			
			_connection.Disconnect(rsr);
		}
		
		/// <summary>
		/// Method run when the bot is registered to the IRC server.
		/// </summary>
		protected void OnRegistered()
		{
			// Stop Identd, no need for it anymore.
			Identd.Stop();
			Log.Success("Identd", "Stopped service daemon");
			Log.Success("Alaris", "Bot registered on server");
			
			// join channels here
			
			foreach(string chan in _channels)
			{
				if(Rfc2812Util.IsValidChannelName(chan))
					_connection.Sender.Join(chan);
				
				Log.Notice("Alaris", "Joined channel: " + chan);
			}
			
			_manager.RunRegisteredHandlers();
			
		}
		
		/// <summary>
		/// Method run when the bot receives a CTCP request.
		/// </summary>
		/// <param name="command">
		/// The CTCP command.
		/// </param>
		/// <param name="user">
		/// Data about the user who sent it.
		/// </param>
		protected void OnCtcpRequest(string command, UserInfo user)
		{
			Log.Notice("CTCP", "Received command " + command + " from " + user.Nick);
		}
	}
	

}